feat(api): (2/n) evaluation + grader api

yanxi0830 opened the pull request:
# PR Stack
- https://github.com/meta-llama/llama-stack/pull/1664
- https://github.com/meta-llama/llama-stack/pull/1691
- https://github.com/meta-llama/llama-stack/pull/1700
- https://github.com/meta-llama/llama-stack/pull/1766

# What does this PR do?
- Simplify evaluation related API using `/evaluation` + `/graders`. These sets of APIs will replace `eval` & `scoring` related APIs. 
- Still keep existing endpoints until full migration. 

- Previous discussions:
  - https://github.com/meta-llama/llama-stack/pull/1580
  - https://github.com/meta-llama/llama-stack/pull/1614

[//]: # (If resolving an issue, uncomment and update the line below)
[//]: # (Closes #[issue-number])

## Client SDK Usage
```python
# to submit an evaluation job
eval_job = client.evaluation.run(
    task={
        "type": "benchmark_id",
        "benchmark_id": "my-benchmark"
    },
    candidate=...
)

# to run evaluation inline
eval_response = client.evaluation.run_sync(
    task={
        "type": "data_source_grader",
        "data_source": {
                "type": "rows",
                "rows": [
                        {
                                  "messages": [{"role": "user", "content": "hello"}], 
                                  "answer": "hello",
                        }
                 ]
        },
        "grader_ids": ["equality"],
    },
    candidate=...,
)

# to submit a grading job with rows
grade_job = client.evaluation.grade(
    task={
        "type": "data_source_grader",
        "data_source": {
                "type": "rows",
                "rows": [
                        {
                                  "messages": [{"role": "user", "content": "hello"}], 
                                  "generation": "hello, what can I do for you",
                                  "answer": "hello",
                        }
                 ]
        },
        "grader_ids": ["equality"],
    },
)
```

## Test Plan


**evaluation.run**
<img width="1274" alt="image" src="https://github.com/user-attachments/assets/5500fafc-c850-474e-87a5-7462512c4304" />

**evaluation.grade_inline**
<img width="1280" alt="image" src="https://github.com/user-attachments/assets/1ba9d8c7-5397-4577-99aa-a2dd041c5ae8" />

**graders**
<img width="1271" alt="image" src="https://github.com/user-attachments/assets/30530aaa-cbd7-4d46-9f6e-80006e40f8eb" />

**benchmarks**
<img width="1279" alt="image" src="https://github.com/user-attachments/assets/afd56d31-3b7c-42a2-84b8-6c763a296801" />

[//]: # (## Documentation)

yanxi0830 committed:
datasets api


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

yanxi0830 committed:
precommit

yanxi0830 committed:
openapi

yanxi0830 committed:
remove json_schema_type decorator

yanxi0830 committed:
update

yanxi0830 committed:
precommit

yanxi0830 committed:
comment

yanxi0830 committed:
upgrade doc

yanxi0830 committed:
Merge branch 'main' into pr1573

yanxi0830 committed:
Merge branch 'main' into pr1573

yanxi0830 committed:
Merge branch 'main' into pr1573

yanxi0830 committed:
comments

yanxi0830 committed:
update

yanxi0830 committed:
source

yanxi0830 committed:
source

yanxi0830 committed:
huggingface obey consistency

yanxi0830 committed:
update doc

yanxi0830 committed:
update doc

yanxi0830 committed:
update doc

yanxi0830 committed:
update doc

yanxi0830 committed:
more purposes

yanxi0830 committed:
docs

yanxi0830 committed:
Merge branch 'main' into pr1573

yanxi0830 committed:
datasetio->datasets

yanxi0830 committed:
docs

yanxi0830 committed:
feat(dataset api): (1.1/n) dataset api implementation fix pre-commit (#1625)

# What does this PR do?
- fix pre-commit with api updates

[//]: # (If resolving an issue, uncomment and update the line below)
[//]: # (Closes #[issue-number])

## Test Plan
```
pre-commit
```

[//]: # (## Documentation)

yanxi0830 committed:
Merge branch 'main' into pr1573

yanxi0830 committed:
Merge branch 'main' into pr1573

yanxi0830 committed:
Merge branch 'main' into pr1573

yanxi0830 committed:
feat(api): (1.2/n) datasets.iterrorws pagination api updates (#1656)

# What does this PR do?
- as title
- uses "cursor" pagination scheme for iterrows

[//]: # (If resolving an issue, uncomment and update the line below)
[//]: # (Closes #[issue-number])

## Test Plan
<img width="1226" alt="image"
src="https://github.com/user-attachments/assets/3220eaac-7117-4d0a-b344-2bbb77a22065"
/>


[//]: # (## Documentation)

yanxi0830 committed:
openapi gen + precommit fix

yanxi0830 committed:
feat(datasets api): (1.3/n) patch OpenAPI gen for datasetio->datasets (#1657)

# What does this PR do?
- We need to tag DatasetIO class correctly with Datasets with the
endpoint change

[//]: # (If resolving an issue, uncomment and update the line below)
[//]: # (Closes #[issue-number])

## Test Plan
**Before**
<img width="1474" alt="image"
src="https://github.com/user-attachments/assets/48737317-28a3-4aa6-a1b5-e1ea680cef84"
/>


**After**
<img width="1508" alt="image"
src="https://github.com/user-attachments/assets/123322f0-a52f-47ee-99a7-ecc66c1b09ec"
/>

[//]: # (## Documentation)

yanxi0830 committed:
feat(dataset api): (1.4/n) fix resolver signature mismatch (#1658)

# What does this PR do?
- fix datasets api signature mis-match so that llama stack run can start

[//]: # (If resolving an issue, uncomment and update the line below)
[//]: # (Closes #[issue-number])

## Test Plan
```
llama stack run
```
<img width="626" alt="image"
src="https://github.com/user-attachments/assets/59072d1a-ccb6-453a-80e8-d87419896c41"
/>


[//]: # (## Documentation)

yanxi0830 committed:
feat(dataset api): (1.5/n) fix dataset registeration (#1659)

# What does this PR do?

- fix dataset registeration & iterrows
> NOTE: the URL endpoint is changed to datasetio due to flaky path
routing

[//]: # (If resolving an issue, uncomment and update the line below)
[//]: # (Closes #[issue-number])

## Test Plan
```
LLAMA_STACK_CONFIG=fireworks pytest -v tests/integration/datasets/test_datasets.py
```
<img width="854" alt="image"
src="https://github.com/user-attachments/assets/0168b352-1c5a-48d1-8e9a-93141d418e54"
/>


[//]: # (## Documentation)

yanxi0830 committed:
fix hf url endpoint

yanxi0830 committed:
scoring fix

yanxi0830 committed:
pre

yanxi0830 committed:
feat(dataset api): (1.6/n) fix all iterrows callsites (#1660)

# What does this PR do?
- as title

[//]: # (If resolving an issue, uncomment and update the line below)
[//]: # (Closes #[issue-number])

## Test Plan
CI

```
pytest -v -s --nbval-lax ./docs/notebooks/Llama_Stack_Benchmark_Evals.ipynb
```
<img width="587" alt="image"
src="https://github.com/user-attachments/assets/4a25f493-501e-43f4-9836-d9802223a93a"
/>


[//]: # (## Documentation)

yanxi0830 committed:
fix integeration

yanxi0830 committed:
precommit

yanxi0830 committed:
dataaset

yanxi0830 committed:
grader

yanxi0830 committed:
new apis

yanxi0830 committed:
precommit

yanxi0830 committed:
benchmarks

raghotham commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,175 @@
+        ...
+
+    @webmethod(route="/evaluation/run_inline", method="POST")
+    async def run_inline(
```
run_sync?

yanxi0830 replied:
done

yanxi0830 committed:
inline -> sync

yanxi0830 committed:
pre

ashwinb commented on llama_stack/apis/resource.py:
```diff
@@ -14,6 +14,8 @@ class ResourceType(Enum):
     vector_db = "vector_db"
     dataset = "dataset"
+    grader = "grader"
+    # TODO: migrate scoring_function -> grader
```
can we add deprecation warnings into scoring_functions as part of this PR?

yanxi0830 replied:
Let's keep this PR for API only for now, and add deprecation later with implementation fix.

ashwinb commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,175 @@
+        candidate: EvaluationCandidate,
+    ) -> EvaluationResponse:
+        """
+        Run an evaluation job inline.
```
explicitly contrast against `run` above, saying:

> Run an evaluation synchronously, i.e., without scheduling a job". You should use this for quick testing, or when the number of rows is limited. Some implementations may have stricter restrictions on inputs which will be accepted.

booxter replied:
Why is this needed as a separate endpoint at all? Can't the client just run and then wait for completion?

yanxi0830 replied:
done

ashwinb replied:
@booxter it is very common to iterate on evaluations with various combinations of (input prompts, grader parameterizations) -- in our experience at Meta it has turned out again and again that scheduler overhead disincentivizes ML researchers from even exploring. so from a DX perspective, no this is not at all equivalent.

ashwinb commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,175 @@
+    @webmethod(route="/evaluation/grade", method="POST")
+    async def grade(self, task: EvaluationTask) -> EvaluationJob:
+        """
+        Run an grading job with generated results. Use this when you have generated results from inference in a dataset.
```
nit: `a grading`. Also we should use the word "response" or "model response".

ashwinb commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,175 @@
+    @webmethod(route="/evaluation/grade_sync", method="POST")
+    async def grade_sync(self, task: EvaluationTask) -> EvaluationResponse:
+        """
+        Run an grading job with generated results inline.
```
same comments as above

yanxi0830 replied:
done

ashwinb commented on llama_stack/apis/common/job_types.py:
```diff
@@ -3,21 +3,49 @@
+    cancelled = "cancelled"
+
+
+class JobArtifact(BaseModel):
```
where is this referenced?

yanxi0830 replied:
removed

ashwinb commented on llama_stack/apis/benchmarks/benchmarks.py:
```diff
@@ -45,22 +51,39 @@ class ListBenchmarksResponse(BaseModel):
-    ) -> None: ...
+    ) -> Benchmark:
+        """
+        Register a new benchmark.
```
I think you should explain what a benchmark is and when you should think of this term

yanxi0830 replied:
done

booxter reviewed (commented):
I'm very confused. Why are previous PRs relevant to the topic closed and new opened? Why are these PRs posted for private branches? The previous discussion focused on jobs behavior was here: https://github.com/meta-llama/llama-stack/pull/1614 and there were comments that I don't think are addressed here. (This PR actually looks like a complete shift in a different direction.)

---

I am also confused about the proposed API and the fact that we seem to entangle different topics - jobs management and eval - in the same PR. Since jobs behavior is going to be inherited for different flows, I'd rather look at it in a separate context.

ehhuang commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,175 @@
+
+        :param task: The task to evaluate. One of:
+         - BenchmarkTask: Run evaluation task against a benchmark_id
+         - DatasetGraderTask: Run evaluation task against a dataset_id and a list of grader_ids
+         - DataSourceGraderTask: Run evaluation task against a data source (e.g. rows, uri, etc.) and a list of grader_ids
```
Can we combine these two? To a evaluation user, this is a small enough difference imo.

yanxi0830 replied:
The flow is a little bit different: 

- `DatasetGraderTask ` requires a dataset.register() to get a dataset_id
- `DataSourceGraderTask` skips the dataset registration flow. 

We can combine it on the SDK side I feel? Otherwise if in the future we decide to skip `datasets`, we'll need to re-update this API.

yanxi0830 replied:
addressed

ehhuang commented on llama_stack/apis/eval/eval.py:
```diff
@@ -91,7 +91,7 @@ async def run_eval(
         self,
```
is this being replaced by /evaluation/benchmarks?

yanxi0830 replied:
No, this will be deleted completely.

ehhuang commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,175 @@
+    :param system_message: (Optional) The system message providing instructions or context to the model.
+    """
+
+    type: Literal["model"] = "model"
```
n00b q: when do we use a enum for type vs not?

yanxi0830 replied:
I guess all type should be an enum for consistency, will update

yanxi0830 replied:
keeping as Literal["model"]

ehhuang commented on llama_stack/apis/graders/graders.py:
```diff
@@ -0,0 +1,237 @@
+
+    grader_type: GraderType
+    description: str
+    supported_dataset_purposes: List[DatasetPurpose] = Field(
```
AFAICT GraderTypeInfo is used by list_grader_type. Where is `supported_dataset_purposes` specified? I don't see it in register_grader.

yanxi0830 replied:
The GraderTypeInfo will be hardcoded on server.

ehhuang commented on llama_stack/apis/graders/graders.py:
```diff
@@ -0,0 +1,237 @@
+    A type of grader. Each type is a criteria for evaluating answers.
+    """
+
+    llm = "llm"
+    regex_parser = "regex_parser"
+    equality = "equality"
+    subset_of = "subset_of"
+    factuality = "factuality"
+    faithfulness = "faithfulness"
```
add documentation for each of these?

yanxi0830 replied:
done

ehhuang commented on llama_stack/apis/graders/graders.py:
```diff
@@ -0,0 +1,237 @@
+    :cvar accuracy: Number of correct results over total results.
+    """
+
+    average = "average"
+    median = "median"
+    categorical_count = "categorical_count"
+    accuracy = "accuracy"
```
Do users ever want just one or a subset of these? Is it too much to always return these?

yanxi0830 replied:
Sometime score are not numerical and we cannot return all aggregation functions.

yanxi0830 replied:
yeah, then we don't need the aggregation type?

ehhuang replied:
Were you going to remove this?

yanxi0830 replied:
yeah, I'll remove once I finish the chunk of GradeType migration.

ehhuang commented on llama_stack/apis/graders/graders.py:
```diff
@@ -0,0 +1,237 @@
+    A type of aggregation function.
+    :cvar average: Average the scores of each row.
+    :cvar median: Median the scores of each row.
+    :cvar categorical_count: Count the number of rows that match each category.
```
What's the output of this? a dict?

yanxi0830 replied:
Yes, a dict, e.g. {"A": 10, "B": 12}

ehhuang commented on llama_stack/apis/graders/graders.py:
```diff
@@ -0,0 +1,237 @@
+    :cvar average: Average the scores of each row.
+    :cvar median: Median the scores of each row.
+    :cvar categorical_count: Count the number of rows that match each category.
+    :cvar accuracy: Number of correct results over total results.
```
is this the same as `sum`?

yanxi0830 replied:
accuracy will have: {num_correct: 10, num_total: 20, accuracy: 0.5}.

ehhuang replied:
(Can we document this?)

yanxi0830 commented on llama_stack/apis/benchmarks/benchmarks.py:
```diff
@@ -45,22 +51,39 @@ class ListBenchmarksResponse(BaseModel):
+        ...
+
+    @webmethod(route="/benchmarks", method="GET")
+    async def list_benchmarks(self) -> ListBenchmarksResponse:
```
Which APIs have List[type]? I remember we made an effort such that all list request will be paginated with a `data` field.

ashwinb replied:
why `list` vs. `List`? typically in our codebase we have stuck to `typing.List[...]` 

re: some APIs using `list[type]`, we should have all of them switch to a wrapper object which has `.data` as the payload which represents the list. That structure allows for proper pagination, should it be needed.

booxter replied:
I think you have to use `List` for older Pythons. But not anymore.

booxter replied:
One example of this is `list_runtime_tools`.

yanxi0830 replied:
fixed

yanxi0830 commented:
@booxter I want to make it clear that this PR do not address Job related changes, and only focuses on Evaluation & Grader APIs. 

The jobs API is not fixed and only added as a temporary change to unblock evaluation & grader. We can focus addressing job related APIs in a separate PR.

ashwinb commented on llama_stack/apis/evaluation/__init__.py:
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
I think Xi said this above but only very short -- but the desire is to replace:
-  '/eval' and '/scoring' -> '/evaluation'
- '/scoring_functions' -> '/graders'

Re: why both? Only for backwards compatibility for a couple releases if folks are already using them somewhere.

yanxi0830 replied:
added back  eval / scoring / scoring_functions

yanxi0830 commented on llama_stack/apis/graders/graders.py:
```diff
@@ -0,0 +1,237 @@
+        return self.provider_resource_id
+
+
+class GraderInput(CommonGraderFields, BaseModel):
```
Let's file an issue to address this across the codebase for all APIs?

yanxi0830 commented on llama_stack/apis/graders/graders.py:
```diff
@@ -0,0 +1,237 @@
+        return self.provider_resource_id
+
+
+class GraderInput(CommonGraderFields, BaseModel):
```
This is part of convention on handling Resources.

yanxi0830 commented on llama_stack/apis/graders/graders.py:
```diff
@@ -0,0 +1,237 @@
+    provider_grader_id: Optional[str] = None
+
+
+class ListGradersResponse(BaseModel):
```
b/c for list responses, we want to have a `data` field to better support pagination. See models.list()

yanxi0830 replied:
1) Yeah, I'll update list_graders (my miss). For list_runtime_tools, it's outside the scope of this PR. 

2) Here's an example OpenAPI spec with pagination. Note that we don't implement the "next" yet, but just wraping it in `data` is standard convention for list responses and we can easily add pagination in the future. 

<img width="495" alt="image" src="https://github.com/user-attachments/assets/21d4b0ba-4dae-43d7-a10c-dd4bd51ca3d1" />

hardikjshah commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,175 @@
+
+
+@json_schema_type
+class DatasetGraderTask(BaseModel):
```
Can we just use `benchmark`, `dataset` and `rows` ? 
`data_source_grader` is really confusing my brain 

```
EvaluationTask = register_schema(
    Annotated[
        Union[EvalBenchmark, EvalDataset, EvalRows],
        Field(discriminator="type"),
    ],
    name="EvaluationTask",
)
@json_schema_type
class EvalBenchmark(BaseModel):
    type: Literal["benchmark"] = "benchmark"
    benchmark_id: str

@json_schema_type
class EvalDataset(BaseModel):
    type: Literal["dataset"] = "dataset"
    dataset_id: str
    grader_ids: List[str]

@json_schema_type
class EvalRows(BaseModel):
    type: Literal["rows"] = "rows"
    data_source: DataSource
    grader_ids: List[str]

yanxi0830 replied:
The DataSource could also have "URI" in addition to "rows".

ehhuang replied:
+1 Made a similar comment below to simplify. In my mind there's just (data, graders) and benchmark (an alias for (data, graders)).

```
@json_schema_type
class EvaluationTask(BaseModel):
    type: Literal["dataset"] = "dataset"
    data: Union[DATASET_ID, DataSource],
    grader_ids: List[str]

@json_schema_type
class BenchmarkEvaluationTask(BaseModel):
    type: Literal["benchmark"] = "benchmark"
    benchmark_id: str
```

yanxi0830 replied:
chatted with @hardikjshah offline, API wise expose, SDK could wrap around

```
class BenchmarkEvaluationTask(BaseModel)
   type: "benchmark"
   benchmark

class DatasetEvaluationTask
   type: "dataset"
   dataset_id:
   grader_ids;

class DataEvaluationTask
    type: "data"
    data_source: DataSource
    grader_ids
```

yanxi0830 replied:
done

yanxi0830 replied:
@ashwinb suggested a better option

```python
@json_schema_type
class EvaluationTask(BaseModel):
    """
    A task for evaluation. To specify a task, one of the following must be provided:
    - `benchmark_id`: Run evaluation task against a benchmark_id
    - `dataset_id` and `grader_ids`: Run evaluation task against a dataset_id and a list of grader_ids
    - `data_source` and `grader_ids`: Run evaluation task against a data source (e.g. rows, uri, etc.) and a list of grader_ids

    :param benchmark_id: The benchmark ID to evaluate.
    :param dataset_id: The dataset ID to evaluate.
    :param data_source: The data source to evaluate.
    :param grader_ids: The grader IDs to evaluate.
    """

    benchmark_id: Optional[str] = None
    dataset_id: Optional[str] = None
    data_source: Optional[DataSource] = None
    grader_ids: Optional[List[str]] = None
```

yanxi0830 commented on llama_stack/distribution/stack.py:
```diff
@@ -17,16 +17,15 @@
 from llama_stack.apis.benchmarks import Benchmarks
 from llama_stack.apis.datasetio import DatasetIO
 from llama_stack.apis.datasets import Datasets
-from llama_stack.apis.eval import Eval
```
The files are present because it is difficult to have a gigantic PR with removals + API updates for review. Implementation will be addressed separately.

yanxi0830 commented on llama_stack/apis/common/job_types.py:
```diff
@@ -3,21 +3,49 @@
+    A job artifact is a file or directory that is produced by a job.
+    """
+
+    path: str
```
outside scope of this pr, adding TODO

yanxi0830 committed:
Merge branch 'main' into eval_api_final

yanxi0830 committed:
precommit

yanxi0830 committed:
revert job related

booxter commented on llama_stack/apis/common/job_types.py:
```diff
@@ -3,7 +3,9 @@
 # the root directory of this source tree.
+from datetime import datetime
 from enum import Enum
+from typing import Optional
```
We should stop adding Optional imports because lately Python support this natively with `type | None` construct.

yanxi0830 replied:
done

booxter commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,175 @@
+
+
+class Evaluation(Protocol):
+    @webmethod(route="/evaluation/run", method="POST")
```
In some other proposals (training?), I've seen jobs created with a POST to `/<API>/` or `/<API>/job/`. Here it's `run`. We should have consistency across these endpoints.

yanxi0830 replied:
We have decided not to have `/job` for job creation. cc @raghotham  @ashwinb

booxter replied:
@yanxi0830 OK. So then we'll have `/run` for all APIs?

yanxi0830 replied:
We will have: 

`/evaluation/run` --> full evaluation job including generation
`/evaluation/grade` --> evaluation job without generation
`finetuning/sft` --> sft finetune job

ashwinb replied:
@booxter we went through many naming iterations and eventually landed on sacrificing a bit of consistency for the sake of user clarity. "running an evaluation" made sense. otherwise, we'd have to invent another verb and we weren't happy with other choices. we tried a "job", but that made it awkward with other considerations.

yanxi0830 committed:
Merge branch 'main' into eval_api_final

yanxi0830 committed:
comments

yanxi0830 committed:
address

yanxi0830 commented:
> Why is this needed as a separate endpoint at all? Can't the client just run and then wait for completion?
@booxter This is for ease of running with inline library client & debugging on just running evaluation for a few dataset rows without querying job status & getting job results.

yanxi0830 commented on llama_stack/apis/graders/graders.py:
```diff
@@ -0,0 +1,237 @@
+        ...
+
+    @webmethod(route="/graders", method="GET")
+    async def list_graders(self) -> List[Grader]:
```
changed to ListGraderResponse

yanxi0830 commented on llama_stack/apis/graders/graders.py:
```diff
@@ -0,0 +1,237 @@
+@runtime_checkable
+class Graders(Protocol):
+    @webmethod(route="/graders", method="POST")
+    async def register_grader(
```
added unregister

yanxi0830 commented on llama_stack/apis/common/job_types.py:
```diff
@@ -3,21 +3,49 @@
+    id: str
+    status: JobStatus
+    created_at: datetime
+    ended_at: Optional[datetime] = None
```
Done, but note that this PR is not used for defining jobs API.

yanxi0830 commented on llama_stack/apis/eval/eval.py:
```diff
@@ -91,7 +91,7 @@ async def run_eval(
         benchmark_id: str,
         benchmark_config: BenchmarkConfig,
-    ) -> Job:
+    ) -> CommonJobFields:
```
done

yanxi0830 commented on llama_stack/apis/benchmarks/benchmarks.py:
```diff
@@ -45,22 +51,39 @@ class ListBenchmarksResponse(BaseModel):
-
-    @webmethod(route="/eval/benchmarks", method="POST")
+    @webmethod(route="/benchmarks", method="POST")
     async def register_benchmark(
```
added unregister

yanxi0830 commented on llama_stack/apis/common/job_types.py:
```diff
@@ -3,21 +3,49 @@
+
+    id: str
+    status: JobStatus
+    created_at: datetime
```
added, note this is not the final jobs api

yanxi0830 commented on llama_stack/apis/benchmarks/benchmarks.py:
```diff
@@ -45,22 +51,39 @@ class ListBenchmarksResponse(BaseModel):
+        """
+        Register a new benchmark.
+
+        :param dataset_id: The ID of the dataset to used to run the benchmark.
```
done

yanxi0830 committed:
comments

booxter commented:
RE `_sync` version of the eval call. I understand how it would be used; I'd argue we don't need to have it and the logic can be easily implemented on client side if needed. Keeping API surface smaller seems to me like a good goal.

yanxi0830 commented:
> RE `_sync` version of the eval call. I understand how it would be used; I'd argue we don't need to have it and the logic can be easily implemented on client side if needed. Keeping API surface smaller seems to me like a good goal.

@booxter Thanks for the careful review. 

I think this behaviour cannot be easily implemented on the client side if we do not want to schedule a job for running eval?

yanxi0830 committed:
pre

ashwinb commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,175 @@
+    """
+
+    type: Literal["agent"] = "agent"
+    config: AgentConfig
```
I think this should be either `agent:` or `agent_config:`

yanxi0830 replied:
updated to `agent_config`

yanxi0830 commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,175 @@
+        candidate: EvaluationCandidate,
+    ) -> EvaluationJob:
+        """
+        Run an evaluation job.
```
done

yanxi0830 committed:
docs

yanxi0830 committed:
grading

yanxi0830 committed:
agent config

yanxi0830 committed:
update

ashwinb commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,177 @@
+        """
+        ...
+
+    @webmethod(route="/evaluation/grading", method="POST")
```
should this route be `/grade` and not `/grading`?

yanxi0830 replied:
ah yes

yanxi0830 committed:
update EvaluationTask

yanxi0830 committed:
update EvaluationTask

ashwinb commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,168 @@
+class EvaluationTask(BaseModel):
+    """
+    A task for evaluation. To specify a task, one of the following must be provided:
+    - `benchmark_id`: Run evaluation task against a benchmark_id
```
```suggestion
    - `benchmark_id`: Run evaluation task against a benchmark_id. Use this when you have a curated dataset and have settled on the graders.
```

ashwinb commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,168 @@
+    """
+    A task for evaluation. To specify a task, one of the following must be provided:
+    - `benchmark_id`: Run evaluation task against a benchmark_id
+    - `dataset_id` and `grader_ids`: Run evaluation task against a dataset_id and a list of grader_ids
```
```suggestion
    - `dataset_id` and `grader_ids`: Run evaluation task against a dataset_id and a list of grader_ids. Use this when you have datasets and / or are iterating on your graders. 
```

ashwinb commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,168 @@
+    A task for evaluation. To specify a task, one of the following must be provided:
+    - `benchmark_id`: Run evaluation task against a benchmark_id
+    - `dataset_id` and `grader_ids`: Run evaluation task against a dataset_id and a list of grader_ids
+    - `data_source` and `grader_ids`: Run evaluation task against a data source (e.g. rows, uri, etc.) and a list of grader_ids
```
```suggestion
    - `data_source` and `grader_ids`: Run evaluation task against a data source (e.g. rows, uri, etc.) and a list of grader_ids. Prefer this when you are early in your evaluation cycle and experimenting much more with your data and graders.
```

yanxi0830 committed:
Update llama_stack/apis/evaluation/evaluation.py

Co-authored-by: Ashwin Bharambe <ashwin.bharambe@gmail.com>

yanxi0830 committed:
Update llama_stack/apis/evaluation/evaluation.py

Co-authored-by: Ashwin Bharambe <ashwin.bharambe@gmail.com>

yanxi0830 committed:
Update llama_stack/apis/evaluation/evaluation.py

Co-authored-by: Ashwin Bharambe <ashwin.bharambe@gmail.com>

yanxi0830 committed:
Merge branch 'main' into eval_api_final

yanxi0830 committed:
openapi gen

yanxi0830 committed:
precommit

yanxi0830 committed:
feat(eval api): (2.1/n) fix resolver for benchmark routing table + fix precommit (#1691)

# What does this PR do?
- fixes routing table so that `llama stack run` works
- fixes pre-commit
- one of many fixes to address implementation fix

[//]: # (If resolving an issue, uncomment and update the line below)
[//]: # (Closes #[issue-number])

## Test Plan
```
llama stack run
```

[//]: # (## Documentation)

ehhuang commented on llama_stack/apis/benchmarks/benchmarks.py:
```diff
@@ -45,22 +51,46 @@ class ListBenchmarksResponse(BaseModel):
+        """
+        Register a new benchmark. A benchmark consists of a dataset id and a list of grader ids.
+
+        :param dataset_id: The ID of the dataset to be used to run the benchmark.
+        :param grader_ids: List of grader ids to use for this benchmark.
+        :param benchmark_id: (Optional) The ID of the benchmark to register. If not provided, an ID will be generated.
```
It would be useful to document how these ids are obtained, e.g. use client...register_x(), even though it's a bit redundant.

yanxi0830 replied:
done.

ehhuang commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,168 @@
+    """
+
+    generations: List[Dict[str, Any]]
+    scores: Dict[str, ScoringResult]
```
Why not just `grades`? so it's consistent with `graders`

yanxi0830 replied:
removed

ehhuang commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,168 @@
+
+
+@json_schema_type
+class EvaluationResponse(BaseModel):
```
One use case I keep coming back to is that the generations within an eval job will fail. When that happens, how do we enable users to easily resume from the previous incomplete job? Do you think this is within scope of this PR? 
Previously we had mentioned returning the dataset ID that includes the inputs and generations. Does that still apply? We also need to include the grades in such dataset as well (so we don't regrade unnecessarily).

ehhuang replied:
One option might be:
```
result_dataset: List[Dict[str, Any]]  # includes original data + generations + grades
grades: List[Dict[str, Any]] # subset of the above with grades only (can also just omit since it's redundant)
aggregated_grades: ...
```
Then user can directly pass `result_dataset` to another `grade()` call, if we implement the corresponding behavior in `grade` to ignore already generated responses/grades.

yanxi0830 replied:
> One use case I keep coming back to is that the generations within an eval job will fail. When that happens, how do we enable users to easily resume from the previous incomplete job?

I don't think we should add the ability to resume from eval runs right now. We should just return back a job error, and make user trigger the job again. It is also hard to do in a production environment, we would need a persistent storage somewhere, along with authentications, job metadata etc. 

> Previously we had mentioned returning the dataset ID that includes the inputs and generations. Does that still apply?

This is easy to do with adding common JobArtifact (see previous discussion https://github.com/meta-llama/llama-stack/pull/1614), but that's outside scope of this PR. 

> We also need to include the grades in such dataset as well (so we don't regrade unnecessarily).
I think the purpose of `EvaluationResponse` is only for returning back some results when executing with `_sync` for debugging.

With jobs, these eval results will just be stored in a dataset file, returned back in JobArtifact.

yanxi0830 replied:
Updated to `result_data` (for dataset rows) & `metrics` (for aggregated results). This feels more aligned with results from actual eval run.

ehhuang replied:
> I don't think we should add the ability to resume from eval runs right now.

Agreed, but ideally have the API able to support that with minimal change.

> I think the purpose of EvaluationResponse is only for returning back some results when executing with _sync for debugging.

> With jobs, these eval results will just be stored in a dataset file, returned back in JobArtifact.

The actual returned results should be consistent, i.e. the columns/rows should be the same across both _sync and non-_sync, right?

> Updated to result_data (for dataset rows) & metrics (for aggregated results).

Or `result_rows`? `metrics` is another term very similar in meaning to `grades` and doesn't convey that they are aggregations, hence I prefer `aggregated_grades` or something more explicit.

yanxi0830 committed:
result_data in evaluation response

ehhuang commented on llama_stack/apis/evaluation/evaluation.py:
```diff
@@ -0,0 +1,155 @@
+    """
+    A response to an inline evaluation.
+
+    :param result_data: The result data containing generations and grades in each row.
```
I think we should include input data here too. Otherwise, users will need to join this with their input data to look at the samples, which is almost guaranteed to happen. Might as well save them that step.

yanxi0830 replied:
added doc.

yanxi0830 committed:
update

yanxi0830 committed:
update

ehhuang reviewed (commented):
One last question. LGTM otherwise!

yanxi0830 committed:
feat(eval api): (2.2/n) delete eval / scoring / scoring_fn apis (#1700)

# What does this PR do?
- To make it easier, delete existing `eval/scoring/scoring_function`
apis. There will be a bunch of broken impls here. The sequence is:
1. migrate benchmark graders
2. clean up existing scoring functions

- Add a skeleton evaluation impl to make tests pass. 

## Test Plan
tested in following PRs

[//]: # (## Documentation)

yanxi0830 committed:
merge

yanxi0830 committed:
remove ScoringFn

yanxi0830 committed:
precommit

yanxi0830 committed:
precommit

yanxi0830 committed:
remove aggregation functions

yanxi0830 committed:
precommit

yanxi0830 committed:
precommit

yanxi0830 committed:
fix precommit

yanxi0830 committed:
fix precommit

yanxi0830 committed:
feat(eval api): (2.3/n) remove scoring / eval impls + benchmarks (#1766)

# What does this PR do?
- Remove `/eval` and `/scoring` impls
- Clean up benchmarks. The benchmarks exists in the `llama-stack-evals`
repo.
- Rest of grading functions will be added in follow up PR. 

[//]: # (If resolving an issue, uncomment and update the line below)
[//]: # (Closes #[issue-number])

## Test Plan
- CI

[//]: # (## Documentation)

dmartinol commented:
will the `Llama_Stack_Benchmark_Evals.ipynb` notebook be part of another PR?

yanxi0830 commented:
> will the `Llama_Stack_Benchmark_Evals.ipynb` notebook be part of another PR?

@dmartinol yes, Llama_Stack_Benchmark_Evals.ipynb update will be follow up work in this big PR.

yanxi0830 committed:
Merge branch 'main' into eval_api_final

yanxi0830 committed:
distro codegen

yanxi0830 committed:
openapi
